### PR TITLE
Changed dependency to the require, not require-dev

### PR DESCRIPTION
**Pull request by danhanly**
egeloen/http-adapter is required for this module to function.
I've added it to the full require, not the require-dev

### CODE REVIEW

**Comment by sagikazarmark** [comment]
You are wrong. Ivory HTTP Adapter is not required, but suggested. It is an optional dependency for the default implementation.

The reason why we chose this path is that you can implement your custom clients as well. I know this can be confusing. Maybe the HTTP transport could go into a separate package?

Also, I noticed that the PHP version requirement is added to the `require-dev` section which is IMO wrong.

/cc @lstrojny

**Comment by danhanly** [comment]
My issue with that, is that I included `lstrojny/fxmlrpc` as a dependency to my project, initialised the client _without any arguments_, and I was faced with the following error:

```
PHP Fatal error:  Class 'Ivory\HttpAdapter\HttpAdapterFactory' not found in htdocs/vendor/lstrojny/fxmlrpc/src/fXmlRpc/Client.php on line 75
```

I should, by all accounts, be able to use the package 'out of the box' without errors. Optional packages should be separated from the main installation, so that anybody who chooses not to use them, isn't hindered by errors.

I think it'll have to be as you suggested; separate the HTTP Transport from module.

**Comment by sagikazarmark** [comment]
Actually it works out of the box if you install the necessary dependencies.

Ivory HTTP adapter is the same: if there is no HTTP Clients installed manually (not even curl extension) then it won't work.

The reason why you get an error is that Client tries to load a http adapter if no parameters are passed to the constructor. The alternative is to remove this feature and always require the developer to define the transport. The result is the same: client accepts an instance of TransportInterface. Since currently only the HttpTransport is included in the core, you would still have to install that dependency manually, because it is not a hard dependency of the client.

**Comment by danhanly** [comment]
If it's giving me an error then the guess method of the HttpAdapterFactory, is effectively guessing the wrong one.

Perhaps then, consider adding the simplest and most lightweight adapter as a dependency, thereby selecting that in the event of a missing parameter? Alternatively, force the selection of an adapter, i.e. don't make it an optional parameter.

`XmlReaderParser` is a default for the `$parser` argument, and `XmlWriterSerializer` is a default for the `$serializer`, argument. However, there's no included default for the `$transport` one.

**Comment by sagikazarmark** [comment]
There is. That's the point. The default is the HttpAdapterTransport. You just have to install the dep. How is it any different than making the transport a required argument and still having the adapter package as a suggestion?

It is just not a hard dependency. Maybe we could mention it in the readme.

**Comment by lstrojny** [comment]
We could also throw an exception that mentions installing the dependency. Opinions?

**Comment by sagikazarmark** [comment]
I am reluctant to accept it as a real solution. It would mix code and package dependencies up, which is IMO not a good idea.

I suggest considering one of the following proposals:

- Make HttpAdapter a requirement in the transport (not the package). (Less progressive solution)
- An extension of the previous: Add a client factory which uses the HttpAdapterFactory similar to the current case. That way an error is only thrown if someone uses the factory, not the client. (Using the factory is optional)
- Separate the HTTP Adapter transport into another package (This makes me think if the transport layer is still necessary. We could make the core package the contract without implementation)

**Comment by sagikazarmark** [comment]
This is going to be solved in the new organization. fXmlRpc (and PHP HTTP replaces ivory) will use virtual packages thus it will be your responsibility to install a transport and an http client.

**Comment by frederikbosch** [comment]
@sagikazarmark @lstrojny @danhanly The opener of this issue is right. Version 0.9.5 is not backward compatible with 0.9.4. Before 0.9.5 one was not required to pass an implementation of ```TransportInterface``` to the constructor of fXmlRpc\Client. Since 0.9.5 one must require Ivory or pass another implementation of TransportInterface to the constructor. Hence, the API has actually changed.

Semantic versioning is not completely clear on libraries that are still in the 0.x range, but I think it would have been a much better idea to release 0.9.5 as 0.10.0.

**Comment by frederikbosch** [comment]
Moreover: it is not even possible anymore to use other implementations of ```TransportInterface``` because they have been deleted from the library in 0.9.5. In my opinion that release is breaking a lot of implementations which is completely unnecessary.

**Comment by lstrojny** [comment]
@frederikbosch indeed, 0.9.5 breaks more then it should. The reason for that is that we are standardizing on the Ivory HTTP adapter that solves the HTTP adapter compatibility issue much better then we ever would. As the change only affects instantiation, my hope is that the breakage is acceptable as it only happens on instantiation time and that’s something you probably do in a single place only. Again, sorry for the hassle. If it is too much work right now, I am very willing to release a 0.9.4.1 with just the bugfix included.

**Comment by frederikbosch** [comment]
@lstrojny No problem. I am actually very happy with the library. In this case I only do not understand why you chose to release it as 0.9.5 instead of 0.10.0. In my case, I was happy to find out quickly and change my composer reference from ~0.9.2 to 0.9.4. This gives time to change my implementation.

Honestly, you do not have to release a 0.9.4.1 for me as I have just fixed the problem. And I also think it will not help people that used the minor version reference, say ~0.9.0, like I did. Because composer then still picks up 0.9.5. Maybe, you could add a BC break warning for 0.9.5 in the top of the readme for a while.

**Comment by tomsommer** [comment]
You really need to release a bugfix release for 0.9 and then break BC in 0.10 if that is what you want. Please follow SemVer.

**Comment by frederikbosch** [comment]
@tomsommer I think that would be the best solution in this case too. To make everybody happy, the release should be the same as 0.9.4 + ivory adapter (but not as default). Then people that already jumped to 0.9.5 would still be happy (I guess).

**Comment by sagikazarmark** [comment]
Guys, I am planning to release the new fxmlrpc in the next few weeks (3-4).
I don't think that semver really matters here, since this package is going
to be abandoned.

**Comment by frederikbosch** [comment]
@sagikazarmark Sounds great! But are you semver does not matter here? Already three people that hit the bug.

**Comment by sagikazarmark** [comment]
I meant you can workaround the problem by installing a specific commit. So
that semver doesn't play a role until the new package is out. But maybe I
am wrong. The point is rather prepare for the new package than care about
some semver problem which only exists because a new release has been tagged
at all.

**Comment by frederikbosch** [comment]
@sagikazarmark Semver always plays a role when external people are leaning on/using packages: that is exactly why it exists. The fact there is a new fxmlrpc coming up is very nice but not relevant in this case. This may sound rude, and I before I say it I want to repeat that I am very happy with the package, but you are now actually saying is: to hell with the current group of users.

Because that is the case. The following people that are hit by the bug.
* People that were not passing an implementation of ```TransportInterface```
* People that were using one of transporters that existed before 0.9.5
with the requirement that they are running ```composer update``` with  ```~``` or ```^``` or ```*```. And my guess is that that group represents your complete user group.

**Comment by sagikazarmark** [comment]
@frederikbosch I am sorry, maybe I wasn't completely clear. It was never my intention to suggest I don't care about user needs.

> Semver always plays a role when external people are leaning on/using your package: that is exactly why it exists.

Of course, you are right. AFAIK the following happened in order:

- Ivory transport has been added added and a lot of refactoring has been done
- A specific bug has been fixed (which required the release, #44)

Which means it is impossible to tag a release which includes the fixed bug AND does not break BC (without going back in history and rewritting it). This is why I said semver doesn't play a role here, because the problem can only be solved with a lot of work to make everyone happy (referring to your suggestion). So what I would do is simply deleting the release and releasing `0.10`. Not everyone will be happy with it, but solves the semver issue.

Another thing you should consider related to semver: this version of fxmlrpc has never reached 1.0, thus it cannot be treated as stable. While it was stable enough, MJZ releases suggest: "You can use it, but watch out, the next release might contain BC breaks" Of course by next release I mean a MINOR one, so it still doesn't mean that a BC break should be released as a PATCH release. However this answers my other comment that another package is coming out soon. While you are free to use non-stable packages, you should be aware of the fact that it's API might not be stable and might be changed. So everyone using it should be able to either do nothing and use a specific version with probably less features or watch the development closely and update the code with a new release.

To sum up: I agree that this release should never have been existed because it caused all problems. Someone said once that Semver is about communication: how you communicate your changes to everyone else. So 0.10 should be a message that we are working on the package and we have made changes that will not be compatible with your code.

I hope I made myself clear and addressed all the issues that were caused by misunderstanding me.

**Comment by frederikbosch** [comment]
@sagikazarmark Thank you for your reply. I agree with the majority of your points, and on the remaining points discussion seems unnecessary to me. Looking forward to new fXmlRpc releases anyway!

**Comment by tomsommer** [comment]
You should release 0.10.0 with the BC and release 0.9.6, which would be the same as 0.9.4

Thanks :)

**Comment by frederikbosch** [comment]
@sagikazarmark You really should consider the idea of @tomsommer. Because #46 is another person who is running against the bug.

**Comment by sagikazarmark** [comment]
Actually, I am fine with both solutions and it is not my call anyway.

Good to see many people using this library (even if it turned out thanks to an unfortunate issue).

Speaking of versions: I am planning to release the new package starting from 0.1 again. I don't think it takes much time to finalize things, just to be able to have previews before the final stable. I am a little bit conerned it might cause confusion.

Ping @lstrojny

**Comment by tomsommer** [comment]
Don't do that. Just continue the version bumps (according to SemVer) :)

**Comment by lstrojny** [comment]
@frederikbosch, @tomsommer thank you guys for your input, the 0.10.0 + 0.9.6 idea makes the most sense to me. I am sorry we fucked up the release process so badly, fXmlRpc is supposed to be infrastructure code that "just works" or at least breaks foreseeably.

**Comment by lstrojny** [comment]
Released 0.10.0 and 0.9.6

**Comment by sagikazarmark** [comment]
@tomsommer So should it be 0.11 for example?

**Comment by lstrojny** [comment]
To help with upgrading, I added a couple of sentences on how to upgrade from 0.9.x to 0.10.x: https://github.com/lstrojny/fxmlrpc#upgrading-to-010x

**Comment by frederikbosch** [comment]
@lstrojny  :+1: Great work! Do not worry on 'fucking' up. This is an awesome library!

**Comment by frederikbosch** [comment]
@lstrojny And I guess this PR can be closed! :)

**Comment by bigwhoop** [comment]
Hi,

I know this issue is closed but I still wanted to share some insight. I'm currently looking into PHP XML-RPC libraries and came across this library on Packagist. I was reading the README and saw the Change Log with lots of breaking changes on minor versions. The lack of semver made me search the issues; that's how I landed here.

For the company I work for/me the absence of semver sadly is a deal-breaker and we'll not be using this library just for this reason. It's a bummer because the library looks really good. I don't want to get into a discussion about the pros/cons of semver, I'd just like to point out that the major frameworks and library vendors in PHP userland seem to have adopted it and that it has had a stabilizing impact on PHP software development. So I'm hoping that this library will adopt it as well some day.

Thanks for your work
Phil

**Comment by lstrojny** [comment]
@bigwhoop thanks for sharing your concerns.

To understand the scope of the breaking changes so far, we made them in areas that are either quite unlikely in use but still technically are a B/C break (an example is the change in the `ParserInterface` which would only be in used if somebody implemented there own parser), could be fixed by simple search and replace (we e.g. changed the namespace from `FXMLRPC` to `fXmlRpc`) or where affecting only the instantiation of the client (and therefore in most applications fixable by changing a factory or a DI container configuration).

To address the point about semver specifically. The semver standard allows us to change basically everything as long as we are before 1.0. 

> 4. Major version zero (0.y.z) is for initial development. Anything may change at any time. The public API should not be considered stable.

What I can tell you as that we will continue to be careful with BC breaks and it’s in my best interest: we use fXmlRpc in production now for years and anything else than easily fixable changes would be a waste of time.

If this is too vague for you, you could consider buying insurance against future BC breaks by simply wrapping the client in your own component and only using this component in your application.
